content/4_Gold/Ternary_Search.mdx

```diff
@@ -0,0 +1,287 @@
+prerequisites:
+  - binary-search
+description: 'Using ternary or binary search to find the mode of unimodal functions.'
+frequency: 1
+---
+
+<Resources>
+	<Resource
+		source="CP-Algorithms"
+		title="Ternary Search"
+		url="https://cp-algorithms.com/num_methods/ternary_search.html"
+		starred
+	/>
+</Resources>
+
+## Introduction
+
+In competitive programming, we are often asked to find the minimum or maximum value we can attain under some conditions. More specifically,
+we are often asked to formulate some function $f(x)$, and find the maximum or minimum value of $f(x)$.
+
+If our function happens to be **unimodal**, then we can use either
+ternary search or binary search to find the mode of our function
+efficiently.
+
+### Definitions
+
+A function is unimodal if:
+1. The function **strictly increases** first, then reaches a maximum, then **strictly decreases**.
+2. The function **strictly decreases** first, then reaches a minimum, then **strictly increases**.
+
+Let's assume that we want to find the local minimum of our function. Then, a
+function is convex if **consecutive differences are non-decreasing**. In calculus
+terms, if the derivative of the function is non-decreasing, the function is convex.
+
+Note that this implies that a convex function is also a unimodal function.
+
+## Ternary Search
+
+For now, let's assume that we want to find the minimum of our function. Given that
+our function is unimodal, this means that the function **strictly decreases** first,
+reaches its minimum, then **strictly increases**.
+
+The idea behind ternary search is somewhat similar to binary search. At every
+step of our algorithm, we want to cut out some large portion of our search
+space from being considered.
+
+Let's consider two points $m_1$ and $m_2$, where $l < m_1 < m_2 < r$. Here,
+$l$ and $r$ are the endpoints of our current search space. This ends up dividing
+our search space into three sections.
+
+Now, we just do a bit of casework to see if we can eliminate any sections.
+
+1. $f(m_1) > f(m_2)$ means that we can eliminate $[l, m_1)$
+2. $f(m_1) < f(m_2)$ means that we can eliminate $(m_2, r]$
+3. $f(m_1) = f(m_2)$ means that we can eliminate $[l, m_1)$ and $(m_2, r]$
+
+<!--
+insert manim anim later
+-->
+
+Typically, we treat case 3 as being analogous to cases 1 and 2.
+
+As a mini-proof for why the way we handle case 1 is correct, consider the
+following:
+
+> If $f(m_1) > f(m_2)$, then we know the following two facts:
+> -  $f(m_1)$ is not the minimum
+> - $f(m_1)$ cannot be on the strictly increasing side of the function
+> Regarding the second point, if $f(m_1) > f(m_2)$ and $m_1$ was on the strictly increasing side of the function, that would imply $m_1 > m_2$, which is a contradiction.
+>
+> Thus, we can conclude that $m_1$ remains on the strictly decreasing section of our function. Because of this, we can conclude that our mode will not be in the range $[l, m_1)$, so we can eliminate this section from our search space.
```

Comment:
can you use minimum / maximum instead of mode when applicable? As mode can have multiple definitions it introduces confusion.